elasticsearch/sandboxes/heka/output/elasticsearch_bulk_api.lua

```diff
@@ -144,6 +148,37 @@ end
 
 batch_count = 0
 retry       = false
+local try_count = 0
+
+local function send_batch2()
+    try_count = try_count + 1
+    local ret, err = send_batch()
+    if ret < 0 then client = nil end -- always use a new connection after an error
+
+    if ret == 0 then
+        retry = false
+    elseif discard and try_count > maxretry then
+        retry = false
+        err = "Discarding "..batch_count.." messages after "..try_count.." attemps, ret="..ret.." err="..err
+        ret = -1
+    elseif abort and try_count > maxretry then
+        retry = true -- not saved by preserve_data for now
+        error("Abort sending "..batch_count.." messages after "..try_count.." attemps, ret="..ret.." err="..err)
```

Comment:
See `shutdown_on_terminate`
https://github.com/mozilla-services/hindsight/blob/master/docs/configuration.md#default-sandbox-configuration-variable

They are saved when a plugin is cleanly stopped and are restored when the plugin is initialized before starting.